### ISSUE TITLE
HDF5 dataset value order

### ISSUE THREAD

**Issue by mpokorny** [StanfordLegion/legion#538]
I've written a test program that creates a small 2-D (3x2) LogicalRegion, initializes it, writes it to HDF5, reads it back from the file into a new LogicalRegion, and compares the two regions. All of this works as I would expect. However, looking at the HDF5 file using h5dump, I find that the elements in the dataset are not in the correct order; in fact, they appear to have reversed row and column indexes. If I lie and provide reversed dimensions when creating the HDF5 dataspace but leave the rest of the code unchanged, the file contains the expected values but at reversed indexes. This effect makes it inconvenient to use the HDF5 outside of Legion. If you want a small reproducer, I can provide it.

**Comment by streichler** [comment]
Yes, this is expected behavior.  Memory instances in Legion can use either Fortran-style array layouts (first index changes fastest) or C-style (last index changes fastest), but default for Fortran order.  HDF5 only supports C-style ordering of data.  Legion allows you to select the order at the time you attach to the external HDF5 instance, but we'll need @elliottslaughter to tell us if/how you control that from Regent code.

**Comment by elliottslaughter** [comment]
CC @magnatelee

**Comment by elliottslaughter** [comment]
Isn't there an argument to be made here that HDF5 instances should follow the logical rather than the physical ordering of data? Otherwise mapping decisions would be capable of causing your code to be incorrect.

**Comment by mpokorny** [comment]
FWIW, I'm using the C++ API. I don't see that AttachLauncher::attach_hdf5() has an argument to select the memory instance order -- should I be setting the launcher's `constraint` value?
Following on the last comment by @elliottslaughter, I would like to see HDF5 files written from Legion to be usable outside of Legion, without any additional (non-libhdf5) annotation of the layout that other applications would have to understand.

**Comment by streichler** [comment]
Oops, I was looking at the `column_major` field on the `attach_array_aos` call right below the `attach_hdf5` call you're using...  I think we will need some new plumbing in Realm (and then in Legion) for this, so I'll take ownership of this for now.

**Comment by mpokorny** [comment]
Thanks for your attention to this issue. It's not a top priority issue for me at the moment as long as a write/read cycle to/from HDF5 within Legion maintains consistency, which appears to be the case currently. Perhaps there are cases that might not work at present (e.g, HDF5 file chunking?), but I won't worry about that until I hit one of those.

**Comment by streichler** [comment]
Commit 8ea7b99 adds control of the dimension order of an external HDF5 instance to the Realm API.  We now need @lightsighter to hook it up to the layout constraints that come from the mapper.  Note that for HDF5 instances, the fastest-moving dimension is the last one (e.g. `2` for a 3-D instance), so an ordering of X, Y, Z would be `dim_order = { 2, 1, 0 }`.

**Comment by lightsighter** [comment]
I've added support for this with 83fd2ebad. I still think Realm should be consistent in the meaning of it's 'dim_order' parameters for instance creation and do any translations internally, but I've done the right thing for Legion for translating layout constraints to Realm's current interface. 

I noticed when doing this that we also don't have support for specifying dimension order (or extents for that matter) for normal posix files either, which is something that is trivial for me to add from the Legion side. @streichler do we want to fix this in Realm too?
